web/pandas/pdeps/0013-standardize-apply.md

```diff
@@ -0,0 +1,158 @@
+# PDEP-13: Deprecate the apply method on Series and DataFrame and make the agg and transform methods operate on series data
+
+- Created: 24 August 2023
+- Status: Under discussion
+- Discussion: [#52140](https://github.com/pandas-dev/pandas/issues/52509)
+- Author: [Terji Petersen](https://github.com/topper-123)
+- Revision: 2
+
+## Abstract
+
+The `apply`, `transform` and `agg` methods have very complex behavior when given callables because they in some cases operate on elements in series, in some cases on series and sometimes try one first, and it that fails, falls back to try the other. There is not a logical system how these behaviors are arranged and it can therefore be difficult for users to understand these methods.
+
+It is proposed that `apply`, `transform` and `agg` in the future will work as follows:
+
+1. the `agg` and `transform` methods of `Series`, `DataFrame` and `groupby` will always operate series-wise and never element-wise
+2. `Series.apply` and `DataFrame.apply` will be deprecated.
+3. The current behavior when supplying string to the methods will not be changed.
+4. `groupby.apply` will not be deprecated (because it behaves differently than `Series.apply` and `DataFrame.apply`)
+
+The above changes means that the future behavior, when users want to apply arbitrary callables in pandas, can be described as follows:
+
+1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.
```

Comment:
Are we sure this is accurate? For the use cases I've seen of `DataFrame.apply`, I think `DataFrame.agg` would probably be more likely as a replacement

Currently, say I do
```python
In [43]: df = pd.DataFrame({'a': ['quetzal', 'quetzal', 'baboon'], 'b': ['panda', 'chinchilla', 'elk']})

In [44]: df
Out[44]:
         a           b
0  quetzal       panda
1  quetzal  chinchilla
2   baboon         elk

In [45]: def func(row):
    ...:     return f"{row['a']} or {row['b']}?"
    ...:

In [46]: df.apply(func, axis=1)
Out[46]:
0         quetzal or panda?
1    quetzal or chinchilla?
2            baboon or elk?
dtype: object
```

Then if I use `DataFrame.map`, it'll operate over all elements of the dataframe 1-by-1. Whereas if I use `.agg`, then I'll preserve what I'm currently doing:
```python
In [47]: df.agg(func, axis=1)
Out[47]:
0         quetzal or panda?
1    quetzal or chinchilla?
2            baboon or elk?
dtype: object
```

Reply:
Isn't this a transform? But users don't get to control what gets passed to the UDF they supply - is it columm by column, row by row, or the entire frame. Perhaps they should.

Reply:
transform would fail here:
```python
In [53]: df.transform(func, axis=1)
---------------------------------------------------------------------------

ValueError: Function did not transform
```

And this underscores just how confusing this is

> is it columm by column, row by row, or the entire frame. Perhaps they should.

yes, exactly, it could be:
`axis=None`: element by element (current behaviour)
`axis=1`: row by row
`axis=0`: element by element

Then, `df.apply(lambda row: ..., axis=1)` can just become `df.map(lambda row: ..., axis=1)`

This is a very common way of using `apply`, e.g.
```python
titanic_df['Person'] = titanic_df[['Age','Sex']].apply(get_person,axis=1)
```
from https://www.kaggle.com/code/omarelgabry/a-journey-through-titanic

Reply:
@MarcoGorelli , your example is an aggregation though and `func` does operate on rows, not elements, so everything works as intended in your `agg` example.

However, I guess `agg` linguistically(?) is best understood as a numerical reduction, not a text combination operation, so using `agg` may be a bit unintuitive. My original proposal would have kept `apply` for situations like this, because the problem I wanted to solve was   `apply` giving subtly different results for `Series.apply` and `DataFrame.apply` (including for `axis=1`).

Your example could be an argument for keeping `apply`, but drop the element-by-element behavior of `Series.apply`, i.e. revert to the original proposal of this PDEP.

I'm -1 on making `map` operate on rows/columns. It's a strength of the method that it  always gives same-shaped results and it would be confusing for the result changing shape based on if an axis parameter is given or not.

Reply:
true, it's a horizontal aggregation: it combines multiple columns into 1

I'm just thinking about the upgrade process

I've only ever seen people use `DataFrame.apply` to create a single column, and if the upgrade process is going to be:
- Series.apply(func) => Series.map(func)
- DataFrame.apply(func, axis=1) => DataFrame.agg(func, axis=1)

then that could be confusing, and certainly harder than just grepping for `.apply`

Reply:
wait let's not add yet another method to the API 😅

Reply:
(just another voice here to support the notion that requiring `agg` for this use case feels very unnatural)